ophyd/indexed_device.py

```diff
@@ -0,0 +1,273 @@
+    if isinstance(ranges, (range, str)):
+        return [list(ranges)]
+
+    return [list(ri) for ri in ranges]
+
+
+def _apply_range_to_value(ranges, value):
+    'Get all values from a format string and the given range(s)'
+    return [
+        value.format(*item)
+        for item in itertools.product(*ranges)
+    ]
+
+
+def _mapping_dict_from_ranges(ranges, attrs, *, tuple_access=True):
+    'Create a dictionary to be used by _IndexedChildLevel'
+    mapping_dict = {}
+
+    for index, attr in zip(itertools.product(*ranges), attrs):
+        d = mapping_dict
+        for index_part in index[:-1]:
+            if index_part not in d:
+                d[index_part] = {}
+            d = d[index_part]
+
+        d[index[-1]] = attr
+
+        if tuple_access:
+            mapping_dict[index] = attr
+
+    return mapping_dict
+
+
+def _verify_with_formatter(formatter, format_str, num_args):
+    'Verify a format string with the given formatter'
+    parsed = list(formatter.parse(format_str))
+    unnamed_fields = [
+        field_name
+        for literal_text, field_name, format_spec, conversion in parsed
+        if field_name == ''
+    ]
+    if len(unnamed_fields) != num_args:
+        raise ValueError(
+            f'Expected a format string that would accept {num_args} args, but '
+            f'got one that takes {len(unnamed_fields)} args in string: '
+            f'{format_str!r}'
+        )
+
+    named_fields = [
+        field_name
+        for literal_text, field_name, format_spec, conversion in parsed
+        if field_name
+    ]
+    if named_fields:
+        named_fields = ', '.join(named_fields)
+        raise ValueError(
+            f'Expected a format string that would accept {num_args} args, but '
+            f'got one that takes named fields {named_fields}: {format_str!r}'
+        )
+
+
+class IndexedComponent(Component):
+    '''
+    A Device component that dynamically creates a sub Device
+
+    Parameters
+    ----------
+    attr : str
+        The attribute format to use
+    suffix:
+        The suffix format to use
+    ranges: str, range, or list of iterables
+        A single range of numbers to expand attr and suffix. For example::
+
+            ranges=range(10)
+
+        Or, a list of ranges or iterables, from which individual indexed
+        components will be created by taking the cartesian product of all
+        items, for example::
+
+            ranges=[range(2), 'AB']
+
+        will be expanded to::
+
+            [(0, 'A'), (0, 'B'), (1, 'A'), (1, 'B')]
+
+        The definition of all attributes to be created is based on the above.
+        Assuming component_class is Component, the following pseudo-code
+        shows what would happen on the Device::
+
+            for attr, suffix in expand_attr_and_suffix(ranges):
+                attr = Component(SignalClass, _suffix, **keyword_arg_dict)
+
+    allow_tuple_access : bool, optional
```

Comment:
Why would you want to dis-allow tuple access?

Might also want to support `dev[0, 'A']`?

Reply:
Supported now with the slicing syntax